[Feat] 커뮤니티 기능 구현

Sw-twt opened the pull request:
📝 변경 사항 요약 (Summary of Changes)

  커뮤니티 기능 구현 (문의, 답변, 리뷰 CRUD)

  📚 관련 이슈 (Related Issues)

  #[이슈번호]
#15 #16 #17 #18 #19 #20 #22 #23 #24 #25 #26 

  ✨ 핵심 변경 사항 (Core Changes)

  1. 문의(Inquiry) 기능

  구현 내용:
  - 상품별 문의 등록/조회/수정/삭제
  - 비밀글 기능 (작성자와 판매자만 조회 가능)
  - 답변 완료 후 수정 불가 처리

  2. 답변(InquiryReply) 기능

  구현 내용:
  - 판매자만 답변 등록 가능 (requireSeller 미들웨어)
  - 답변 등록 시 문의 상태 자동 변경 (WaitingAnswer → CompletedAnswer)
  - 답변 삭제 시 문의 상태 복구 (CompletedAnswer → WaitingAnswer)

  3. 리뷰(Review) 기능

  구현 내용:
  - 구매자만 리뷰 작성 가능 (requireBuyer 미들웨어)
  - 주문 아이템 검증 (실제 구매한 상품만 리뷰 가능)
  - 중복 리뷰 방지
  - 평균 평점 및 리뷰 개수 조회

  4. 인증 미들웨어 추가

  구현 내용:
  커뮤니티 기능을 위한 역할 기반 인증 미들웨어 추가

  - authenticate: JWT 토큰 검증 및 사용자 정보 조회
  - requireBuyer: 구매자 권한 검증
  - requireSeller: 판매자 권한 검증

  주요 파일:
  - src/middlewares/auth.ts (신규 파일)

  참고:
  기존 auth.middleware.ts와 별도로 작성했습니다. 

  5. 기존 인증 미들웨어 수정

  수정 이유:
  커뮤니티 기능 구현 중 TypeScript 빌드 에러 발생
  - req.user 타입이 { id: string; email: string; type: UserRole } 로 정의되어 있으나, 기존 코드는 { id: string } 만 설정하여 타입 불일치 발생

  수정 내용:
  - DB에서 사용자 정보 조회하여 email, type 필드 포함하도록 수정
  - 사용자가 존재하지 않을 경우 에러 처리 추가

  수정 파일:
  - src/middlewares/auth.middleware.ts

  6. 공통 유틸리티 추가

  구현 내용:
  - Custom Error 클래스 (NotFoundError, ForbiddenError, ConflictError, BadRequestError, UnauthorizedError)
  - 통일된 응답 형식 (sendSuccess)
  - TypeScript 타입 정의 (req.user 타입 확장)

  주요 파일:
  - src/utils/errors.ts
  - src/utils/response.ts
  - src/types/express.d.ts


## 🔍 코드 리뷰 시 특별히 더 신경 써줬으면 하는 부분 (Specific areas for review)

## ✅ 체크리스트 (Checklist)

- [x] 빌드 및 테스트를 통과했습니다.
- [x] 코드 스타일 가이드라인을 준수했습니다.
- [x] 리뷰어에게 필요한 모든 정보를 제공했습니다.

Sw-twt committed:
커뮤니티 기능 구현

bgk614 commented:
Prettier 적용 필요
로컬에서 `npm run format` 한 번 실행해주세요!
<img width="437" height="186" alt="image" src="https://github.com/user-attachments/assets/9fb921db-3a43-46e7-87ec-f87fd8a6caa0" />

Lint 적용 필요
`npm run lint:fix` 실행 후에 에러 뜨는 부분은 수동으로 수정 부탁드립니다

bgk614 commented on src/utils/errors.ts:
`src/utils/http-exception.ts` 파일이랑 목적이 겹치는 것 같아요
한 번 확인부탁드립니다 !

bgk614 commented on src/middlewares/auth.ts:
```diff
@@ -0,0 +1,66 @@
+
+  if (req.user.type !== UserRole.SELLER) {
+    throw new ForbiddenError('판매자만 접근할 수 있습니다.');
+  }
```
```suggestion
```
지금 하나의 미들웨어에서 두 개의 책임을 지고 있는 것 같아요

인증 부분은 기존의 auth.middleware.ts 사용하고 
사용자의 권한 체크는 권한 전용 미들웨어를 만드는게 유지보수하기에 더 좋을 것 같아요

Sw-twt committed:
코드 포맷팅 및 린트 에러 수정

Sw-twt committed:
코드 포맷팅 수정

Sw-twt committed:
Merge branch 'dev' into feature/community-api

Sw-twt committed:
포맷팅 수정

Sw-twt committed:
[fix] ESLint import 정렬 및 req.user 타입 정의 수정

ziheon01 reviewed (commented):
먼저 형익님, 커뮤니티 기능 구현하시느라 고생하셨습니다~ 비즈니스 로직은 구현상 잘 작동을 하는 것 같고 손댈곳이 없을 거 같습니다.

다만, 보경님 말씀대로 공통 모듈 부분이 기존 팀 컨벤션과 겹치는 부분이 있어서 팀 표준으로 통합 하면 될 것 같습니다. 

1. auth.ts :
     먼저 authenticate 함수는 기존에 개발된 auth.middleware.ts에 있는 authenticate를 사용하는게 좋을거 같습니다. ( 기기 제한, 기타 로직 포함)
그런 후에 auth.ts  파일은 role.middleware.ts 같은 파일로 이름을 바꾸고 requireBuyer, requireSeller 함수만 남겨서 권한 체크 전용 함수로 사용하는게 더 좋을거 같습니다.

2. 에러 핸들링, 응답 포맷 통일 (errors.ts, response.ts)
이 부분은 기존 팀 컨밴션에 맞게 개발 되어있는 모듈을 사용하는게 맞는 것 같다고 생각합니다 !!

bgk614 commented on src/services/reviewService.ts:
```diff
@@ -0,0 +1,114 @@
+        order: true,
+        product: true,
+      },
+    });
```
```suggestion
```

bgk614 commented on src/services/inquiryService.ts:
```diff
@@ -0,0 +1,105 @@
+  async createInquiry(data: CreateInquiryData) {
+    const product = await prisma.product.findUnique({
+      where: { id: data.productId },
+    });
```
```suggestion
```

bgk614 commented on src/services/inquiryReplyService.ts:
```diff
@@ -0,0 +1,95 @@
+
+    const store = await prisma.store.findUnique({
+      where: { userId },
+    });
```
```suggestion
```

DB 접근은 모두 레포지토리 레이어에서 진행하면 좋을 것 같습니다!

bgk614 commented on src/services/inquiryService.ts:
```diff
@@ -0,0 +1,105 @@
+    } else {
+      const store = await prisma.store.findUnique({
+        where: { userId },
+      });
```
```suggestion
```

bgk614 commented on src/services/inquiryService.ts:
```diff
@@ -0,0 +1,105 @@
+  async getProductInquiries(productId: string) {
+    const product = await prisma.product.findUnique({
+      where: { id: productId },
+    });
```
```suggestion
```

bgk614 commented on src/services/reviewService.ts:
```diff
@@ -0,0 +1,114 @@
+  async getProductReviews(productId: string) {
+    const product = await prisma.product.findUnique({
+      where: { id: productId },
+    });
```
```suggestion
```

bgk614 reviewed (commented):
레포지토리로 옮기면 좋을 것 같은 부분 체크해뒀습니다

Sw-twt committed:
refactor: 커뮤니티 기능 팀 컨벤션 통합

- 에러 처리 통일: HttpException 사용
- 응답 포맷 통일: res.status().json() 직접 사용
- 인증/인가 분리: role.middleware.ts 추가
- 레포지토리 레이어 추가: product, store, orderItem
- 서비스에서 prisma 직접 호출 제거
- API 테스트 파일 추가: community.http

ziheon01 commented:
형익님 먼저 피드백 반영하시느라 고생하셨습니다.

피드백 반영이 아주 잘 돼있는거 같습니다.
사소한 몇 가지만 수정하면 너무 좋은 코드가 될거 같아 리뷰드립니다 !!

service 코드를 보니 에러를 new HttpException으로 직접 생성 하고 계신거 같습니다.
저희 공통 모듈(http-exception.ts)에 두 모듈 추가해주신 후에 서비스 코드에서 사용해주시면 될거같습니다.

```
// 권한 없음 (Forbidden)
  static forbidden(message = MESSAGE.forbidden) {
    return new HttpException({
      status: STATUS_CODE.FORBIDDEN,
      message: message,
    });
  }

  // 리소스 충돌 (Conflict) - 예: 중복 리뷰 등
  static conflict(message: string) {
    return new HttpException({
      status: STATUS_CODE.CONFLICT,
      message: message,
    });
  }
```

마지막으로는 파일 이름 컨벤션도 한번 통일해주면 좋을거같습니다.
ex) camelCase => inqury.controller.ts

일단 머지 진행 후 이슈로 남겨놨다가 진행해도 상관 없을거 같아 approve 진행해놓겠습니다 !!!

Sw-twt committed:
refactor: PR 리뷰 반영
   - 서비스에서 HttpException 정적 메서드 사용
   - 파일명 kebab-case 컨벤션 통일
   - import 경로 업데이트

Sw-twt committed:
Merge branch 'dev' into feature/community-api

Sw-twt committed:
chore: prettier 포맷팅

Sw-twt committed:
fix: user.repository update 메서드 타입 에러 수정

Sw-twt committed:
fix: 사용하지 않는 Prisma import 제거
